### PR TITLE
feat(filesystem): scan in client/server mode

### LINKED ISSUE #634: trivy scan local filesystem not supported in client mode
trivy should support scan local filesystem using client mode
THX

### PR DESCRIPTION
**Pull request by afdesk**
## Description
Added a new option `--server` to `filesystem` command for support local filesystem scan in client/server mode.

Also there was moved `client` package to `artifact/client.go`.

## Related issues
- Close #634 

## Checklist
- [x] I've read the [guidelines for contributing](https://github.com/aquasecurity/trivy/blob/main/CONTRIBUTING.md) to this repository.
- [x] I've followed the [conventions](https://github.com/aquasecurity/trivy/blob/main/CONTRIBUTING.md#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/commands/app.go [hunk 1]
```diff
@@ -313,6 +312,19 @@ var (
 		EnvVars: []string{"TRIVY_INSECURE"},
 	}
 
+	remoteServer = cli.StringFlag{
+		Name:    "remote",
```
[the remark this answers is not in the document]
Comment (afdesk) [reply pkg/commands/app.go]: done

File: pkg/commands/artifact/option.go [hunk 2]
```diff
@@ -18,9 +21,19 @@ type Option struct {
 	option.CacheOption
 	option.ConfigOption
 
+	// For policy downloading
+	NoProgress bool
```
[the remark this answers is not in the document]
Comment (afdesk) [reply pkg/commands/artifact/option.go]: Removed. the comment got confused me.

File: pkg/commands/artifact/option.go [hunk 3]
```diff
@@ -18,9 +21,19 @@ type Option struct {
...
 	// We don't want to allow disabled analyzers to be passed by users,
 	// but it differs depending on scanning modes.
 	DisabledAnalyzers []analyzer.Type
+
+	RemoteAddr    string
+	token         string
+	tokenHeader   string
+	customHeaders []string
+	// this field is populated in Init()
+	CustomHeaders http.Header
```
[the remark this answers is not in the document]
Comment (afdesk) [reply pkg/commands/artifact/option.go]: I've created a new struct: `RemoteOption`

File: pkg/commands/artifact/run.go [hunk 4]
```diff
@@ -92,29 +96,37 @@ func runWithTimeout(ctx context.Context, opt Option, initializeScanner Initializ
 	return nil
 }
 
-func initFSCache(c Option) (cache.Cache, error) {
+func initRemoteCache(c Option) (cache.ArtifactCache, cache.LocalArtifactCache, error) {
+	return remoteCache.NewRemoteCache(c.RemoteAddr, c.CustomHeaders, c.Insecure), nil, nil
+}
+
+func initFSCache(c Option) (cache.ArtifactCache, cache.LocalArtifactCache, error) {
```
[the remark this answers is not in the document]
Comment (afdesk) [reply pkg/commands/artifact/run.go]: I think it's an interesting idea. I've done it

File: pkg/commands/artifact/option_test.go [hunk 5]
```diff
@@ -177,11 +253,17 @@ func TestOption_Init(t *testing.T) {
 			set.Bool("reset", false, "")
 			set.Bool("skip-db-update", false, "")
 			set.Bool("download-db-only", false, "")
+			set.Bool("clear-cache", false, "")
```
Comment (knqyf263) [inline pkg/commands/artifact/option_test.go:256]: Is this needed?

File: pkg/commands/artifact/option_test.go [hunk 6]
```diff
@@ -177,11 +253,17 @@ func TestOption_Init(t *testing.T) {
...
+
 			set.String("severity", "CRITICAL", "")
 			set.String("vuln-type", "os,library", "")
 			set.String("security-checks", "vuln", "")
 			set.String("template", "", "")
 			set.String("format", "", "")
+			set.String("server", "", "")
+			set.String("token", "", "")
+			set.String("token-header", "", "")
+			set.Var(&cli.StringSlice{}, "custom-headers", "")
```
Comment (knqyf263) [inline pkg/commands/artifact/option_test.go:266]: What is this for?

File: pkg/rpc/client/client_test.go [hunk 7]
```diff
@@ -1,116 +1,47 @@
...
-		args            args
-		scanExpectation scanExpectation
-		wantResults     types.Results
-		wantOS          *ftypes.OS
-		wantEosl        bool
-		wantErr         string
+		name          string
+		customHeaders http.Header
+		args          args
+		expectation   *rpc.ScanResponse
```
Comment (knqyf263) [inline pkg/rpc/client/client_test.go:35]: We can use `interface{}` for defining an error.
https://github.com/aquasecurity/trivy/blob/04d86ece69232e64ee9cfca3c6471998fe84d4c5/pkg/rpc/client/client_test.go#L169-L175

File: pkg/commands/option/remote.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+		token:         c.String("token"),
+		tokenHeader:   c.String("token-header"),
+		remote:        c.String("remote"), // deprecated
+	}
+
+	return r
+}
+
+// Init initialize the options for client/server mode
+func (c *RemoteOption) Init(logger *zap.SugaredLogger) (err error) {
```
Comment (knqyf263) [inline pkg/commands/option/remote.go:36]: Looks like we don't need to return an error.

File: integration/client_server_test.go [hunk 9]
```diff
@@ -220,6 +223,15 @@ func TestClientServer(t *testing.T) {
...
 			golden: "testdata/busybox-with-lockfile.json.golden",
 		},
+		{
+			name: "scan pox.xml with fs command in client/server mode",
+			args: csArgs{
+				Command:          "fs",
+				RemoteAddrOption: "--server",
+				Target:           "testdata/fixtures/fs/pom/",
+			},
+			golden: "testdata/fs-pom.json.golder",
```
Comment (knqyf263) [inline integration/client_server_test.go:233]: We have to use the same golden file with standalone mode. Maybe we have some differences now?

File: pkg/commands/artifact/option_test.go (hunk 5 above)
Comment (afdesk) [reply pkg/commands/artifact/option_test.go]: no, removed.

File: pkg/commands/artifact/option_test.go (hunk 6 above)
Comment (afdesk) [reply pkg/commands/artifact/option_test.go:264]: discussed.

File: pkg/commands/option/remote.go (hunk 8 above)
Comment (afdesk) [reply pkg/commands/option/remote.go]: good catch! removed.

File: integration/client_server_test.go (hunk 9 above)
Comment (afdesk) [reply integration/client_server_test.go]: Great idea! done.

File: pkg/rpc/client/client_test.go (hunk 7 above)
Comment (afdesk) [reply pkg/rpc/client/client_test.go:35]: I've changed this block. but i'm not sure that it's correct.

**Comment by knqyf263** [comment]
Thanks!

**Pull request merged**